server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

```diff
@@ -410,6 +410,32 @@ public void complete(ActionListener<?> listener) {
         }
     }
 
+    /**
+     * Creates and returns a new {@link SubscribableListener} {@code L} and subscribes {@code nextStep} to this listener such that if this
+     * listener is completed successfully then the result is discarded and {@code nextStep} is invoked with argument {@code L}. If this
+     * listener is completed with exception {@code E} then so is {@code L}.
+     * <p>
+     * This can be used to construct a sequence of async actions, each ignoring the result of the previous ones:
+     * <pre>
+     * l.andThen(l1 -> forkAction1(args1, l1)).andThen(l2 -> forkAction2(args2, l2)).addListener(finalListener);
+     * </pre>
+     * After creating this chain, completing {@code l} with a successful response will call {@code forkAction1}, which will on completion
+     * call {@code forkAction2}, which will in turn pass its response to {@code finalListener}. A failure of any step will bypass the
+     * remaining steps and ultimately fail {@code finalListener}.
+     * <p>
+     * The threading of the {@code nextStep} callback is the same as for listeners added with {@link #addListener}: if this listener is
+     * already complete then {@code nextStep} is invoked on the thread calling {@link #andThen} and in its thread context, but if this
+     * listener is incomplete then {@code nextStep} is invoked on the completing thread and in its thread context. In other words, if you
+     * want to ensure that {@code nextStep} is invoked using a particular executor, then you must do both of:
```

Comment:
would it make sense to add also the alternative that allows setting executor/context?

Reply:
There's nowhere that would use that alternative today so it'd just be dead code at the moment. Easy enough to add later on if we need it.